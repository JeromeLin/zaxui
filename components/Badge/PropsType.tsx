export default interface PropsType {
  theme?: 'default' | 'primary' | 'info' | 'success' | 'warning' | 'error';
  shape?: 'dot' | 'radius' | 'round' | 'circle';
  sup?: boolean;
  text?: any;
}
