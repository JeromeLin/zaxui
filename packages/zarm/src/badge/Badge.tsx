import * as React from 'react';
import classnames from 'classnames';
import type { BaseBadgeProps } from './interface';

export interface BadgeProps extends BaseBadgeProps, React.HTMLAttributes<HTMLElement> {
  prefixCls?: string;
}

const Badge = React.forwardRef<unknown, BadgeProps>((props, ref) => {
  const { prefixCls, className, theme, shape, text, children, ...restProps } = props;
  const badgeRef = (ref as any) || React.createRef<HTMLElement>();

  const cls = classnames(prefixCls, className, {
    [`${prefixCls}--${theme}`]: !!theme,
    [`${prefixCls}--${shape}`]: shape,
    [`${prefixCls}--sup`]: !!children,
  });

  return (
    <span ref={badgeRef} className={cls} {...restProps}>
      {children}
      <sup className={`${prefixCls}__content`}>{shape !== 'dot' && text}</sup>
    </span>
  );
});

Badge.displayName = 'Badge';

Badge.defaultProps = {
  prefixCls: 'za-badge',
  shape: 'dot',
  theme: 'danger',
};

export default Badge;
