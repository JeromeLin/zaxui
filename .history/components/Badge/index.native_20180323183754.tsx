import React, { PureComponent } from 'react';
import {
  StyleSheet,
  View,
  Text,
} from 'react-native';
import classnames from 'classnames';
// 引入组件参数
import PropsType from './PropsType';
// 引入CSS
import badgeStyle from './style/index.native';

// 问号代表可选
// export default interface PropsType {
//   theme?: 'default' | 'primary' | 'success' | 'warning' | 'error';
//   shape?: 'dot' | 'radius' | 'round' | 'circle';
//   sup?: boolean;
//   text?: any;
// }

export interface BadgeProps extends PropsType {
  style?: CSSProperties;
  styles?: typeof badgeStyle;
}

// 这里的意思是将传入两个参数，React.Component 的参数第一个是Props，第二个是 State，
// 然后利用typescript的类型检查，Props类型需要时上面定义的 BadgeProps中的可选参数中的变量名
// state这里传入任意都行

const badgeStyles = StyleSheet.create<any>(badgeStyle);

export default class Badge extends PureComponent<BadgeProps, {}> {

  static defaultProps = {
    prefixCls: 'za-badge',
    theme: 'error',
    sup: false,
    styles: badgeStyles,
  };

  state = {
    dotWidth: 0,
  };

  // 获取元素宽度设置style
  layout = (e) => {
    let dotWidth = -parseInt(e.layout.width, 10) / 2;
    this.setState({
      dotWidth,
    });
  }

  render() {
    const { styles } = this.props;

    const {
      prefixCls,
      theme,
      shape,
      sup,
      text,
      children,
      ...others,
    } = this.props;

    const bagdeWrapper = [
      styles!.TextStyle,
    ] as ViewStyle;

    const dotText = [
      styles!.dotText,
      styles![`${shape}Text`],
    ];

    const shapeStyle = [
      styles![`${shape}Shape`],
      styles![`${theme}Theme`],
      sup && styles![`${shape}Sup`],
    ] as ViewStyle;

    return (
      <View style={bagdeWrapper}>
        {children}
        <View
          onLayout={({ nativeEvent: e }) => this.layout(e)}
          style={[shapeStyle, { right: sup ? this.state.dotWidth : 0 }]}
        >
          <Text style={dotText}>{text}</Text>
        </View>
      </View>
    );
  }
}
