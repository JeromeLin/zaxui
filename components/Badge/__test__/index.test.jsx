import React from 'react';
import renderer from 'react-test-renderer';
import Badge from '../index';

describe('Badge', () => {
  it('renders shape is dot correctly', () => {
    const component = renderer.create(<Badge shape="dot" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
