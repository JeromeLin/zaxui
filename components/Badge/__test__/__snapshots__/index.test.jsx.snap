// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Badge renders shape is dot correctly 1`] = `
<span
  className="za-badge theme-error shape-dot"
>
  <sup
    className="za-badge-sup"
  />
</span>
`;
